br/pkg/restore/misc.go

```diff
@@ -64,23 +69,56 @@ func GetTableSchema(
 	return table.Meta(), nil
 }
 
-// GetExistedUserDBs get dbs created or modified by users
-func GetExistedUserDBs(dom *domain.Domain) []*model.DBInfo {
+const MAX_USERTABLES_NUM = 10
+
+// AssertUserDBsEmpty check whether user dbs exist in the cluster
+func AssertUserDBsEmpty(dom *domain.Domain) error {
 	databases := dom.InfoSchema().AllSchemas()
-	existedDatabases := make([]*model.DBInfo, 0, 16)
+	ver, err := dom.Store().CurrentVersion(kv.GlobalTxnScope)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	m := meta.NewSnapshotMeta(dom.Store().GetSnapshot(kv.NewVersion(ver.Ver)))
+	userTables := make([]string, 0, MAX_USERTABLES_NUM)
+	appendTables := func(dbName, tableName string) bool {
+		if len(userTables) >= MAX_USERTABLES_NUM {
+			userTables = append(userTables, "...")
+			return true
+		}
+		userTables = append(userTables, fmt.Sprintf("%s.%s", dbName, tableName))
+		return false
+	}
+LISTDBS:
 	for _, db := range databases {
 		dbName := db.Name.L
 		if tidbutil.IsMemOrSysDB(dbName) {
 			continue
-		} else if dbName == "test" && len(db.Tables) == 0 {
+		}
+		tables, err := m.ListSimpleTables(db.ID)
+		if err != nil {
+			return errors.Annotatef(err, "failed to iterator tables of database[id=%d]", db.ID)
+		}
+		if len(tables) == 0 {
 			// tidb create test db on fresh cluster
 			// if it's empty we don't take it as user db
+			if dbName != "test" {
```

Comment:
why do we ignore tables in "test" db?

Reply:
Actually, it records the empty database here. The database `test` has been created since the cluster initialization.